experimental/FTheoryTools/src/WeierstrassModels/attributes.jl

```diff
@@ -242,6 +242,33 @@ julia> length(singular_loci(w))
 2
 ```
 """
+@attr Vector{<:Tuple{<:MPolyIdeal{<:MPolyRingElem}, Tuple{Int64, Int64, Int64}, String}} function singular_loci(w::WeierstrassModel)
+  @req (base_space(w) isa NormalToricVariety || base_space(w) isa FamilyOfSpaces) "Singular loci of Weierstrass model is currently only supported for toric varieties and families of spaces as base space"
+  B = irrelevant_ideal(base_space(w))
+  d_primes = factor(discriminant(w))
+  nontrivial_d_primes = [k for k in d_primes if _is_nontrivial(ideal([k[1]]), B)]
+  f_primes = factor(weierstrass_section_f(w))
+  g_primes = factor(weierstrass_section_g(w))
+  f_prime_dict = Dict(fp[1] => fp[2] for fp in f_primes)
+  g_prime_dict = Dict(gp[1] => gp[2] for gp in g_primes)
+  kodaira_types = Vector{Tuple{<:MPolyIdeal{<:MPolyRingElem}, Tuple{Int64, Int64, Int64}, String}}(undef, length(nontrivial_d_primes))
+  for (i, d_prime) in enumerate(nontrivial_d_primes)
+    f_order = get(f_prime_dict, d_prime[1], 0)
+    g_order = get(g_prime_dict, d_prime[1], 0)
+    d_order = d_prime[2]
+    ords = (f_order, g_order, d_order)
+    kodaira_types[i] = (ideal([d_prime[1]]), ords, _kodaira_type(ideal([d_prime[1]]), ords, w))
+  end
```

Comment:
You can use nested tuples to make this nicer to read:
```suggestion
  for (i, (p, n)) in enumerate(nontrivial_d_primes)
    f_order = get(f_prime_dict, p, 0)
    g_order = get(g_prime_dict, p, 0)
    d_order = n
    ords = (f_order, g_order, d_order)
    I = ideal([p])
    kodaira_types[i] = (I, ords, _kodaira_type(I, ords, w))
  end
```
You could also avoid factoring f and g and do this instead:
```suggestion
  f = weierstrass_section_f(w)
  g = weierstrass_section_g(w)
  for (i, (p, d_order)) in enumerate(nontrivial_d_primes)
    f_order = valuation(f, p)
    g_order = valuation(g, p)
    ords = (f_order, g_order, d_order)
    I = ideal([p])
    kodaira_types[i] = (I, ords, _kodaira_type(I, ords, w))
  end
```